apps/emqx/src/emqx_channel.erl

```diff
@@ -2112,30 +2141,51 @@ maybe_publish_will_msg(
 maybe_publish_will_msg(
     Reason, Channel = #channel{clientinfo = ClientInfo, will_msg = WillMsg, timers = Timers}
 ) when
-    Reason =:= {shutdown, expired} orelse
-        Reason =:= {shutdown, kicked}
+    Reason =:= expired orelse
+        Reason =:= discarded orelse
+        %% Unsure...
+        Reason =:= {shutdown, internal_error} orelse
+        Reason =:= kicked
 ->
-    %% Must publish now without delay and cancel the will message timer.
+    %% For the cases that session MUST be gone.
+    %% a. expired (session expired)
+    %% c. discarded (Session ends because another process starts new session with the same clientid)
+    %% b. kicked. (kicked by operation)
+    %% d. internal_error (maybe not recoverable)
```

Comment:
Agree. It seems a bit unsafe to assume that in the event of hitting internal error publishing should still work.

Reply:
I just list them, may miss something, the entire call path, call stack is hard to follow.

Reply:
I found this in spec:
```
In the case of a Server shutdown or failure, the Server MAY defer publication of Will Messages until a subsequent restart. If this happens, there might be a delay between the time the Server experienced failure and when the Will Message is published.
```
So I think it is ok to publish/not publish/delay publish will when server error.